### ISSUE TITLE
Incremental Adoption

### ISSUE THREAD

**Issue by danieleades** [python/mypy#10413]
This is as much a `pre-commit` question as a `mypy` question.

Mypy is very challenging to apply as a quality gate on large, existing codebases. The very high number of errors prevents using mypy as a quality gate on *all* code, since updating type annotations for the entire codebase in one go is simply not practical.

What i'm finding is that it's difficult to use Mypy for *incremental adoption* of typing in a codebase.

I can use a pre-commit hook to run Mypy on changed files with no issue, but if i'm using the [pre-commit github action](https://github.com/pre-commit/action) in CI, that's going to run on *all* files, not only on *changed* files.

An alternative might be to somehow track the total number of Mypy errors in the codebase, and require that they never *increase* as a result of changes. If there's any way to do this using pre-commit, i've not found any mention of it.

Basically i'm finding Mypy as a continuous integration quality gate to be, in practice, a little "all or nothing".

I guess this is a general problem of incrementally applying quality constraints to legacy code, but i'm finding it to be of particular impact with type annotations and type errors.

Any ideas?

**Comment by danieleades** [comment]
cross-posted here - https://github.com/pre-commit/mirrors-mypy/issues/46
